wallet/key_deriver.go

```diff
@@ -0,0 +1,153 @@
+	counterpartyKey := kd.NormalizeCounterparty(counterparty)
+	invoiceNumber, err := kd.ComputeInvoiceNumber(protocol, keyID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to compute invoice number: %w", err)
+	}
+
+	var pubKey *ec.PublicKey
+	if forSelf {
+		privKey, err := kd.privateKey.DeriveChild(counterpartyKey, invoiceNumber)
+		if err != nil {
+			return nil, fmt.Errorf("failed to derive child private key: %w", err)
+		}
+		pubKey = privKey.PubKey()
+	} else {
+		pubKey, err = counterpartyKey.DeriveChild(kd.privateKey, invoiceNumber)
+		if err != nil {
+			return nil, fmt.Errorf("failed to derive child public key: %w", err)
+		}
+	}
+	return pubKey, nil
+}
+
+func (kd *KeyDeriver) DerivePrivateKey(protocol WalletProtocol, keyID string, counterparty WalletCounterparty) (*ec.PrivateKey, error) {
+	counterpartyKey := kd.NormalizeCounterparty(counterparty)
+	invoiceNumber, err := kd.ComputeInvoiceNumber(protocol, keyID)
+	if err != nil {
+		return nil, err
+	}
+
+	k, err := kd.privateKey.DeriveChild(counterpartyKey, invoiceNumber)
+	if err != nil {
+		return nil, fmt.Errorf("failed to derive child key: %w", err)
+	}
+	return k, nil
+}
+
+func (kd *KeyDeriver) NormalizeCounterparty(counterparty WalletCounterparty) *ec.PublicKey {
+	switch counterparty.Type {
+	case CounterpartyTypeSelf:
+		return kd.privateKey.PubKey()
+	case CounterpartyTypeOther:
+		return counterparty.Counterparty
+	case CounterpartyTypeAnyone:
+		_, pub := ec.PrivateKeyFromBytes([]byte{1})
+		return pub
+	default:
+		return nil
+	}
+}
+
+func (kd *KeyDeriver) ComputeInvoiceNumber(protocol WalletProtocol, keyID string) (string, error) {
+	// Validate protocol security level
+	if protocol.SecurityLevel < 0 || protocol.SecurityLevel > 2 {
+		return "", fmt.Errorf("protocol security level must be 0, 1, or 2")
+	}
+
+	// Validate protocol name
+	protocolName := strings.ToLower(strings.TrimSpace(protocol.Protocol))
+	if len(protocolName) > 400 {
+		// Special handling for specific linkage revelation
+		if strings.HasPrefix(protocolName, "specific linkage revelation ") {
+			if len(protocolName) > 430 {
+				return "", fmt.Errorf("specific linkage revelation protocol names must be 430 characters or less")
+			}
+		} else {
+			return "", fmt.Errorf("protocol names must be 400 characters or less")
+		}
+	}
+	if len(protocolName) < 5 {
+		return "", fmt.Errorf("protocol names must be 5 characters or more")
+	}
+	if strings.Contains(protocolName, "  ") {
+		return "", fmt.Errorf("protocol names cannot contain multiple consecutive spaces (\"  \")")
+	}
+	if !regexp.MustCompile(`^[a-z0-9 ]+$`).MatchString(protocolName) {
+		return "", fmt.Errorf("protocol names can only contain letters, numbers and spaces")
+	}
+	if strings.HasSuffix(protocolName, " protocol") {
+		return "", fmt.Errorf("no need to end your protocol name with \" protocol\"")
+	}
```

Comment:
maybe it's worth to extract a private function for validating protocol name?

this would make the comment above redundant.

Reply:
I'm not opposed. The `ts-sdk` that this was ported from does not abstract that away though. https://github.com/bitcoin-sv/ts-sdk/blob/8e6ed8a/src/wallet/KeyDeriver.ts#L270

Reply:
If this is the only place we need do this, I would prefer we don't abstract it